### DISCUSSION TITLE
参数路由与准确路由同时添加，准确路由有可能被覆盖

### DISCUSSION THREAD

**Discussion by kele1997** [abersheeran/index.py#33] [Q&A] [answered]
```
from indexpy.routing.tree import RadixTree
t = RadixTree()

def donothing():
    pass

def donothing2():
    pass

# 交换下面两个路由的顺序，断言可以通过
t.append("/app/{hello}/", donothing)
t.append("/app/hello/", donothing2)

ret = t.search("/app/world/")
print(ret)   # ({'hello': 'world'}, <function donothing at 0x7f12c94931f0>)
assert ret[1] == donothing

ret = t.search("/app/hello/")
print(ret)  # ({'hello': 'hello'}, <function donothing at 0x7f12c94931f0>)
assert ret[1] == donothing2
```

期望的行为应该类似于 flask 如下：   

```
curl xxxx/a1/hello/ ==> hello2
curl xxxx/a1/world/ ==> hello1
```

```
from flask import Flask

app = Flask(__name__)


@app.route("/a1/<hello>/")
def hello1(hello):
    return "hello1"

@app.route("/a1/hello/")
def hello2():
    return "hello2"

app.run()
```

另外这里 

https://github.com/abersheeran/index.py/blob/e895d06ec26bdb4196f919229417b424f5e066a2/indexpy/routing/tree.py#L57-L66

检查路由冲突的地方

第 60 行的 `   if (node.re_pattern == re_pattern) != (node.characters == param_name):`  会不会有 re_pattern 不同， 而 characters == param_name 相同的情况？(这一块没太看懂）

**Comment by abersheeran** [comment]
https://index-py.aber.sh/stable/route/#_15 这是设计时预料到的情况，并不能算是一个 BUG。

Flask 能够这么匹配，是因为它把精准路由的权重提高了。但 Index.py 没法这么做，因为底层用的 Radix Tree 进行匹配。

虽然不是 BUG，但也许这里可以增加一个运行时输出的 warning，更加友好一些。感谢反馈。

---

检查路由冲突这里，如果 re_pattern 不同，那么就是两条不同的路由（在 radix tree 是两个不同的节点），不会出现歧义（也许也会出现两者同时能匹配的情况，但是这里已经没办法检测到了）。所以只检查了相同 re_pattern 的情况下，name 是不是一样。

**Reply by kele1997** [reply]
golang gin 框架的路由也是基于 radix tree的，它直接禁掉了这种情况，或许直接禁掉是最好的选择。  

但是，其实 index.py 如果先加准确路由，再加参数路由的话，其实可以优先匹配准确路由的。     

（或许可以向 radix tree 的 next_nodes 列表中，在尾部插入参数路由，在头部插入准确路由，然后搜索的时候，按层遍历，每层自左到右遍历，这样或许可以,但是不晓得会不会引入更多的问题....)

**Reply by abersheeran** [reply]
是的，所以我申明了，在同级节点之间，按照加入顺序进行匹配。这样使用者也不需要记住什么优先级问题，很符合人类直觉的，先加入的先匹配。

**Reply by abersheeran** [reply]
禁用也是个好办法，我既然都发 Warning 了，不如直接 raise Error

**Comment by abersheeran** [comment, marked as answer]
https://github.com/abersheeran/index.py/commit/a5ec85af87b7cdc00b598594ddf841b2a0601136 已经提交了检查。
